python/pypto/language/parser/ast_parser.py

```
@@ -494,7 +500,27 @@ def parse_for_loop(self, stmt: ast.For) -> None:
+        # For pl.unroll(), require compile-time constant integer bounds.
+        # Fail early with a clear parser error instead of a later generic
+        # failure in the UnrollLoops C++ pass.
+        if iterator_type == "unroll":
+            for _bound_name in ("start", "stop", "step"):
+                _bound_value = range_args.get(_bound_name)
+                if _bound_value is not None and not isinstance(_bound_value, (int, ir.ConstInt)):
+                    raise ParserSyntaxError(
+                        "pl.unroll() requires compile-time constant integer bounds",
+                        span=self.span_tracker.get_span(iter_call),
+                        hint="Use integer literals for start, stop, and step in pl.unroll().",
+                    )
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)